### DISCUSSION TITLE
Auto-generate the Symphony validation rules from the WDSL/XSD?

### DISCUSSION THREAD

**Discussion by OAFCROB** [phpro/soap-client#358] [Ideas]
### Support Question
Is it possible to auto-generate the Symphony validation rules from the WDSL/XSD?  

The documentation states that you can include the [validator subscriber](https://github.com/phpro/soap-client/blob/master/docs/event-subscribers/validator.md) into your factor method but doesn't give an example of how to define `$validator`. 

I assume that you're supposed to initiate an instance of the Symphony validator and associate the rules to the instance? 

```
public static function factory(string $wsdl): TestClient
{
     $engine = ExtSoapEngineFactory::fromOptions(
            ExtSoapOptions::defaults($wsdl)
                ->withClassMap(TestClassmap::getCollection())

        );

        $eventDispatcher = new EventDispatcher();
        $eventDispatcher->addSubscriber(
            new ValidatorSubscriber(
                Validation::createValidator() // Symphony validator instance 
            )
        );

        return new TestClient($engine, $eventDispatcher);
}
```

What I was wondering though, is it possible to automatically generate the validation rules for the request from the WDSL or the XSD file? 

The WDSL we have is provided to us via a third-party client and we're using this package to make the SOAP calls but ideally, we would like to request exceptions of invalid data format i.e. missing parameters rather than the current SOAP exceptions we're currently getting.

Could you do something like this;

```
$validator = Validation::createValidatorBuilder()
    ->addXmlMapping('some-external-ip/test.xsd')
    ->getValidator();

$eventDispatcher = new EventDispatcher();
        $eventDispatcher->addSubscriber(
            new ValidatorSubscriber(
               $validator
            )
        );
```

**Comment by veewee** [comment]
Hello @OAFCROB,

That sounds like a fun new feature!


First things first: creating a validator can be done in multiple ways, with different kind of configuration files.
All information can be found:
- https://symfony.com/doc/current/components/validator.html
- https://symfony.com/doc/current/components/validator/resources.html

Currently I am using YAML in one of our projects. The validator creating looks like this:

```php
class ValidatorFactory
{
    public static function create(string $configDir): ValidatorInterface
    {
        return Validation::createValidatorBuilder()
            ->addYamlMappings(iterator_to_array(
                Finder::create()
                      ->files()
                      ->in($configDir)
                      ->name('*.yaml')
            ))
            ->getValidator();
    }
}
```

If you don't like yaml, you could always go for annotations or xml. However, it is not possible to provide an XSD and let the validator find out what to do. It is validating PHP objects with validator specific configuration. This will require some xsd to validator configuration transformation.

(The part above is something we could add to the documentation)


About the validator config generation: there are some things we could do... 

First thing you need to know is that we are currently limited by the info provided by ext-soap in this package. We are parsing `__getFunctions()` and `__getTypes()` to build up all the information we need. However, this doesnt contain a lot of advanced xsd information for validation (e.g. min/max occurences, xsd:restrictions, ...)
All the information we have, is available in the `$engine->getMetadata()` method.
Based on this information, you could automatically create validation rules as e.g. annotations on top of the generated types. As mentioned earlier, you won't be able to do advanced validation, since the information is limited.

However, it is possible to create your own `MetadataProviderInterface`. This means you could parse the WSDL and filter out advanced criteria based on the schema (XSD) inside the WSDL. We currently only provide an ext-soap metadata provider. But I was thinking of going ext-soap-less in the future with the help of https://github.com/goetas-webservices/wsdl-reader. A first phase of this could be to only parse metadata based on the wsdl.

As a last possible solution: it might also be possible to create an HTTP middleware that validates the request / response XML based on the XSD in the WSDL (with `DOMDocument::schemaValidate()`). The error messages would not be that great and I am not sure if it would add any value.


Feel free to play around with this! :)

**Comment by OAFCROB** [comment]
@veewee thanks for the reply, it would be super awesome to even have some basic validation auto-generated. Will have to a play around with it and see what we can achieve.

**Comment by WinterSilence** [comment]
You can add mapper tag `@Assert\...` to generator of property comments https://symfony.com/doc/current/components/validator/resources.html#the-annotationloader

**Comment by whatUwant** [comment]
@OAFCROB Does https://github.com/goetas-webservices/xsd2php do the job?

**Comment by OAFCROB** [comment]
@whatUwant that looks interesting, will have to bear that in mind for future projects 😀

**Comment by veewee** [comment]
FYI : we are planning to improve the metadata that can be used to generate code, but it will take some time ... :)
